### PR TITLE
Add bootstrapping support for jenkins users (secure installations).

### PR DESCRIPTION
**Pull request by elconas**
This is another version of https://github.com/jenkinsci/puppet-jenkins/pull/752 that includes support for bootstrapping users via init.groovy.d as this was requested by comments in the issue thread linked to https://github.com/jenkinsci/puppet-jenkins/pull/752 -> https://github.com/jenkinsci/puppet-jenkins/issues/749 ... 

Other issues that may interest this pull are: 
  - https://github.com/jenkinsci/puppet-jenkins/issues/749
  - https://github.com/jenkinsci/puppet-jenkins/issues/671
  - https://github.com/jenkinsci/puppet-jenkins/issues/684
  -  .. maybe others.

### CODE REVIEW

**Comment by oc243** [comment]
👍  for this PR. It fixes the issues that I was having. I'm keen for this to be merged :)

**Comment by egouraud-claranet** [comment]
:+1: 
some useful things inside (for jenkins > 2.46.2 users)

**Comment by elconas** [comment]
Updated the pull request to be in line with the latest master.

**Comment by elconas** [comment]
Generally I would like to see AD and any other wuth support via init.groovy.d, as is makes sure that jenkins is configured properly and puppet code via groovy can function.

**Comment by oc243** [comment]
Would be great to see this one merged ;-)

**Comment by vStone** [comment]
Could you also provide a way to specify the pubkey as a file on the jenkins master? The sshkey is auto generated and so is the pubkey that goes with it :)

File: manifests/users.pp [hunk 1]
```diff
@@ -6,6 +6,28 @@
...
+    if empty($_bootstrap_users) {
+      $_bootstrap_groovy_ensure = absent
+    } else {
+      $_bootstrap_groovy_ensure = present
+    }
+  } else {
+    $_bootstrap_groovy_ensure = absent
+  }
+
+  file { "${::jenkins::jenkins_home}/init.groovy.d/puppet.bootstapping.groovy":
```
Comment (cova-fe) [inline manifests/users.pp]: bootstapping is on purpose or is it a typo?

File: manifests/users.pp [hunk 2]
```diff
@@ -6,6 +6,28 @@
...
+    ensure    => $_bootstrap_groovy_ensure,
+    owner     => $::jenkins::user,
+    group     => $::jenkins::group,
+    mode      => '0640',
+    tag       => 'jenkins_groovy_init_script',
+    show_diff => false,
+    content   => template('jenkins/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb'),
```
Comment (cova-fe) [inline manifests/users.pp]: same as before

**Comment by vStone** [comment]
```diff
diff --git a/templates/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb b/templates/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb
index 8d72d95..e1fcc05 100644
--- a/templates/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb
+++ b/templates/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb
@@ -18,13 +18,18 @@ def currentRealm = instance.getSecurityRealm().getClass().getName()
 
 // bootstrap users
 def bootstrapUsers = [
-<% @_bootstrap_users.each do |username,userconfig| -%>
+<% @_bootstrap_users.each do |username,userconfig|
+-%>
   [
     username : '<%= username %>',
     password : '<%= userconfig['password'] %>',
     email : '<%= userconfig['email'] %>',
     full_name : '<%= userconfig['full_name'] %>',
+<% if userconfig['public_key_file'] -%>
+    ssh_key_file: '<%= userconfig['public_key_file'] %>',
+<% else -%>
     ssh_key : '<%= userconfig['public_key'] %>',
+<% end -%>
   ],
 <% end -%>
 ]
@@ -46,7 +51,13 @@ bootstrapUsers.each { user ->
   def password = user['password']
   def email = user['email']
   def full_name = user['full_name']
-  def ssh_key = user['ssh_key']
+  def ssh_key = null
+  if (user.containsKey('ssh_key_file')) {
+    ssh_key = new File(user['ssh_key_file']).getText('UTF-8')
+  }
+  else {
+    ssh_key = user['ssh_key']
+  }
 
   println "--> Configuring Jenkins Automation User (${username} with ssh_key ${ssh_key})"
 
```

**Comment by vStone** [comment]
@elconas: I added a pr on your fork for my diff: https://github.com/elconas/puppet-jenkins/pull/1

File: templates/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    username : '<%= username %>',
+    password : '<%= userconfig['password'] %>',
+    email : '<%= userconfig['email'] %>',
+    full_name : '<%= userconfig['full_name'] %>',
+    ssh_key : '<%= userconfig['public_key'] %>',
+  ],
+<% end -%>
+]
+
+if (currentRealm == 'hudson.security.HudsonPrivateSecurityRealm') {
```
Comment (cova-fe) [inline templates/home/jenkins/init.groovy.d/puppet.bootstapping.groovy.erb]: If there is a different realm configured, say LDAP, this test will fail and will overwrite the ldap configuration. Shouldn't it be something like `if (currentRealm != 'hudson.security.SecurityRealm$None') {`, more or less?

**Comment by TheKangaroo** [comment]
Any news on that? I would need this to set up secured Jenkins.
I opened a PR (#809) to support pam auth which won't work because of Jenkins limitations (issues.jenkins-ci.org/browse/JENKINS-15063).

**Comment by bastelfreak** [comment]
Hi. Thanks for raising this PR, but I'm going to close it due to inactivity. Please reopen it if you're still interested in getting this merged.

**Pull request closed without merging**